### DISCUSSION TITLE
NFT Image is not displaying on Metamask, even Opensea

### DISCUSSION THREAD

**Discussion by ArnaudBand** [Cyfrin/foundry-full-course-cu#1602] [Q&A] [answered]
### Lesson 11:
I finish deploying my NFT but the problem is not display an image after importing it on Metamsk and even on Opensea. Any help how to fix it

**Comment by EngrPips** [comment]
Hello @ArnaudBand , i ran into an issue where it wasn't displaying on metamask myself and didn't spend too much time to fix it before i moved on but of recently someone else ran into the same problem and they did figure it out and fix it and also they wrote an article on how they fix it on their linkedin profile, maybe that would help you troubleshoot this problem. Here https://www.linkedin.com/pulse/making-your-nft-collection-yevhen-hrytsenko-mjxif/ is the link if you would want to check out what they wrote and how they resolved the issue

**Comment by ArnaudBand** [comment]
Hi @EngrPips, I have an issue at the moment, I changed the info of IPFS using the LinkedIn info, I can't the mint the nft and I can see the the NFT on Opensea.

i got this error when 'm minting

```bash
  [6879] MintDogeNFT::run()
    ├─ [3925] DevOpsTools::get_most_recent_deployment("DOGIE", 11155111 [1.115e7]) [delegatecall]
    │   ├─ [0] VM::readDir("./broadcast", 3) [staticcall]
    │   │   └─ ← [Revert] the path broadcast is not allowed to be accessed for read operations
    │   └─ ← [Revert] the path broadcast is not allowed to be accessed for read operations
    └─ ← [Revert] the path broadcast is not allowed to be accessed for read operations


Error: 
script failed: the path broadcast is not allowed to be accessed for read operations
make: *** [Makefile:74: mint] Error 1

```

**Reply by EngrPips** [reply]
Hello @ArnaudBand, please add the line of code below to your foundry.toml file, and you should stop getting that error. 

`fs_permissions = [{ access = "read", path = "./broadcast" }, {access = "read", path = "./Images/"}]`

**Comment by ArnaudBand** [comment]
Thank you @EngrPips, now the problem, is displaying that I didn't deploy the nft smart contract

```bash


  [451804] MintDogeNFT::run()
    ├─ [448856] DevOpsTools::get_most_recent_deployment("DOGIE", 11155111 [1.115e7]) [delegatecall]
    │   ├─ [0] VM::readDir("./broadcast", 3) [staticcall]
    │   │   └─ ← [Return] [("", "/workspaces/cyfrin_updraft_course/foundry_mood_nft/broadcast/DeployDogeNFT.s.sol", 1, true, false), ("", "/workspaces/cyfrin_updraft_course/foundry_mood_nft/broadcast/DeployDogeNFT.s.sol/11155111", 2, true, false), ("", "/workspaces/cyfrin_updraft_course/foundry_mood_nft/broadcast/DeployDogeNFT.s.sol/11155111/run-1713432503.json", 3, false, false), ("", "/workspaces/cyfrin_updraft_course/foundry_mood_nft/broadcast/DeployDogeNFT.s.sol/11155111/run-latest.json", 3, false, false)]
    │   ├─ [0] VM::toString(11155111 [1.115e7]) [staticcall]
    │   │   └─ ← [Return] "11155111"
    │   ├─ [0] VM::toString(11155111 [1.115e7]) [staticcall]
    │   │   └─ ← [Return] "11155111"
    │   ├─ [0] VM::toString(11155111 [1.115e7]) [staticcall]
    │   │   └─ ← [Return] "11155111"
    │   ├─ [0] VM::toString(11155111 [1.115e7]) [staticcall]
    │   │   └─ ← [Return] "11155111"
    │   ├─ [0] VM::readFile("/workspaces/cyfrin_updraft_course/foundry_mood_nft/broadcast/DeployDogeNFT.s.sol/11155111/run-latest.json") [staticcall]
    │   │   └─ ← [Return] <file>
    │   ├─ [0] VM::parseJsonUint("<stringified JSON>", ".timestamp") [staticcall]
    │   │   └─ ← [Return] 1713432503 [1.713e9]
    │   ├─ [0] console::log("Processing: ", "/workspaces/cyfrin_updraft_course/foundry_mood_nft/broadcast/DeployDogeNFT.s.sol/11155111/run-latest.json") [staticcall]
    │   │   └─ ← [Stop] 
    │   ├─ [0] VM::parseJson("<JSON file>", "$.transactions") [staticcall]
    │   │   └─ ← [Return] <encoded JSON value>
    │   ├─ [0] VM::keyExists("<JSON file>", "$.transactions[0].rpc") [staticcall]
    │   │   └─ ← [Return] false
    │   ├─ [0] console::log("Inspecting %s transactions", 1) [staticcall]
    │   │   └─ ← [Stop] 
    │   └─ ← [Revert] revert: No contract deployed
    └─ ← [Revert] revert: No contract deployed



== Logs ==
  Processing:  /workspaces/cyfrin_updraft_course/foundry_mood_nft/broadcast/DeployDogeNFT.s.sol/11155111/run-latest.json
  Inspecting 1 transactions
Error: 
script failed: revert: No contract deployed

```

**Reply by EngrPips** [reply]
Have you initially deployed a `DogeNft` contract to the sepolia chain? , most likely, you have not. Please open your `broadcast` folder, as shown below, and confirm if you have any `DogeNft` contract deployed initially.

<img width="1438" alt="Screenshot 2024-04-18 at 10 52 45" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/95760918/23184466-bab4-4cec-943e-79cf866c62b8">

**Reply by ArnaudBand** [reply]
I finshed to do that but still, I got the same error. Can we use Discord?

![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/89846108/33c0fcfa-88ff-4f47-ab5c-de9de4228479)

**Reply by EngrPips** [reply]
Open the `11155111` folder instead of `31337`. Sorry for the confusion, and If you prefer Discord, that's fine

**Reply by ArnaudBand** [reply]
@EngrPips, Still it has failed to mint, I saw 2 folders on your broadcast, DeployMoodNFT, MintBasicNFT. What is the content or it's because I don't have them?

![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/89846108/beed938e-caa4-45fd-af7e-1a069344cab1)

![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/89846108/ea5e23e1-3610-41e3-8794-028b5bac17df)

**Reply by EngrPips** [reply]
In the course, we develop a `basic NFT` and a `mood NFT,` and those in the picture I sent are just the folder that contains the deployment details of both of those  `Nft.`

**Reply by EngrPips** [reply]
@ArnaudBand, it seems like there is a conflict between what you name your contract and what name you provided to the `DevOps` tool you are using to get the `most recently deployed contract.`  The picture you shared above shows that the name of your contract is `DogeNFT,` but you pass in `Dogie` as the name of the contract you are searching for. The `DevOps` tool searched for a recent deployment folder that has in it the name of contract `Dogie` but couldn't find hence, it returned `No contract deployed.` 

Please show your `mint` script

**Reply by ArnaudBand** [reply]
I'm not getting you this link is for my repo. Please, can you check the error?
https://github.com/ArnaudBand/cyfrin_updraft_course/tree/main/foundry_mood_nft


```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import {Script} from "forge-std/Script.sol";
import {DevOpsTools} from "lib/foundry-devops/src/DevOpsTools.sol";
import {DogeNFT} from "../src/DogeNFT.sol";

contract MintDogeNFT is Script {
    string public constant PUG_URI =
        "ipfs://bafybeig37ioir76s7mg5oobetncojcm3c3hxasyd4rvid4jqhy4gkaheg4/?filename=0-PUG.json";

    function run() external {
        address mostRecentlyDeployedDogieNft = DevOpsTools.get_most_recent_deployment("DOGIE", block.chainid);
        mintNftOnContract(mostRecentlyDeployedDogieNft);
    }

    function mintNftOnContract(address dogieNftAddress) public {
        vm.startBroadcast();
        DogeNFT(dogieNftAddress).mintNFT(PUG_URI);
        vm.stopBroadcast();
    }
}
```

**Reply by EngrPips** [reply]
Interesting, now show the NFT contract itself

**Reply by ArnaudBand** [reply]
```solidity
// SPDX-Licenser-Identifier: MIT
pragma solidity ^0.8.19;

import "@openzeppelin/contracts/token/ERC721/ERC721.sol";

contract DogeNFT is ERC721 {
    error DogeNFT__TokenUriNotFound();

    mapping(uint256 tokenId => string tokenUri) private s_tokenURIs;
    uint256 private s_tokenCounter;

    constructor() ERC721("DOGIE", "DOG") {
        s_tokenCounter = 0;
    }

    function mintNFT(string memory tokenUri) public {
        s_tokenURIs[s_tokenCounter] = tokenUri;
        _safeMint(msg.sender, s_tokenCounter);
        s_tokenCounter = s_tokenCounter + 1;
    }

    function tokenURI(uint256 tokenId) public view override returns (string memory) {
        if (ownerOf(tokenId) == address(0)) {
            revert DogeNFT__TokenUriNotFound();
        }
        return s_tokenURIs[tokenId];
    }

    function getTokenCounter() public view returns (uint256) {
        return s_tokenCounter;
    }
}
```

**Reply by ArnaudBand** [reply]
Can I call you on discord?

**Reply by EngrPips** [reply, marked as answer]
Now the problem is glaring: the name of your contract in the above picture is `DogeNFT,` but when you used the `DevOps` tool in your mint script, you passed in `DOGIE` as shown in the line of code below. 

` address mostRecentlyDeployedDogieNft = DevOpsTools.get_most_recent_deployment("DOGIE", block.chainid);`

So, change the line in your `mint` script to the below, and your code should now work fine. 

`address mostRecentlyDeployedDogieNft = DevOpsTools.get_most_recent_deployment("DogeNFT", block.chainid);`

**Reply by ArnaudBand** [reply]
Thank you so much @EngrPips, I fix it

![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/89846108/b68817aa-9738-457c-98ed-d3aac4f03947)

**Reply by EngrPips** [reply]
Awesome, I am glad you were able to figure it out. Keep crushing it, friend

**Reply by kinjun751** [reply]
I have the same issue the image doesnt show on metamask

**Reply by EngrPips** [reply]
> I have the same issue the image doesnt show on metamask

Read this discussion https://github.com/Cyfrin/foundry-full-course-cu/discussions/2112
